t/01-pdk/05-client/09-get_protocol.t

```diff
@@ -0,0 +1,314 @@
+    }
+--- request
+GET /t
+--- more_headers
+X-Forwarded-Proto: something other than "https"
+--- response_body
+protocol=http
+--- no_error_log
+[error]
+
+
+
+=== TEST 8: client.get_protocol() returns "http" when the request has no x-forwarded-proto header
+--- http_config eval: $t::Util::HttpConfig
+--- config
+    location = /t {
+        access_by_lua_block {
+            ngx.ctx.route = {
+              protocols = { "http", "https" }
+            }
+
+            local PDK = require "kong.pdk"
+            local pdk = PDK.new()
+            _G.kong = pdk
+            pdk.ip.is_trusted = function() return true end -- mock
+            ngx.say("protocol=", assert(pdk.client.get_protocol(true)))
+        }
+    }
+--- request
+GET /t
+--- response_body
+protocol=http
+--- no_error_log
+[error]
+
+
+
+=== TEST 9: client.get_protocol() returns "tcp" on tcp connections
+--- stream_config eval
+qq{
+    server {
+        listen unix:$ENV{TEST_NGINX_NXSOCK}/nginx.sock proxy_protocol;
+        content_by_lua_block {
+            require "resty.core"
+
+            ngx.ctx.route = {
+              protocols = { "tcp", "tls" }
+            }
+
+            local PDK = require "kong.pdk"
+            local pdk = PDK.new()
+            ngx.say("protocol=", assert(pdk.client.get_protocol()))
+        }
+    }
+}
+--- stream_server_config
+    content_by_lua_block {
+        local sock = ngx.socket.tcp()
+        sock:connect("unix:$TEST_NGINX_NXSOCK/nginx.sock")
+        local request = "PROXY TCP4 10.0.0.1 " ..
+                        ngx.var.server_addr    .. " " ..
+                        ngx.var.remote_port    .. " " ..
+                        ngx.var.server_port    .. "\r\n"
+        sock:send(request)
+        ngx.print(sock:receive())
+    }
+--- stream_response chop
+protocol=tcp
+--- no_error_log
+[error]
+
+
+
+=== TEST 10: client.get_protocol() returns "tls" on tls connections
+--- stream_config eval
+qq{
+    server {
+        listen unix:$ENV{TEST_NGINX_NXSOCK}/nginx.sock proxy_protocol;
+
+        content_by_lua_block {
+            require "resty.core"
+
+            ngx.ctx.route = {
+              protocols = { "tcp", "tls" }
+            }
+
+            ngx.ctx.balancer_data = {
+              scheme = "tls",
+              ssl_ctx = {},
+            }
+
+            local PDK = require "kong.pdk"
+            local pdk = PDK.new()
+            ngx.say("protocol=", assert(pdk.client.get_protocol()))
+        }
+    }
+}
+--- stream_server_config
+    content_by_lua_block {
+        local sock = ngx.socket.tcp()
+        sock:connect("unix:$TEST_NGINX_NXSOCK/nginx.sock")
+        local request = "PROXY TCP4 10.0.0.1 " ..
+                        ngx.var.server_addr    .. " " ..
+                        ngx.var.remote_port    .. " " ..
+                        ngx.var.server_port    .. "\r\n"
+        sock:send(request)
+        ngx.print(sock:receive())
+    }
+--- stream_response chop
+protocol=tls
+--- no_error_log
+[error]
```

Comment:
We are missing test cases ensuring that these APIs (to be called in very hot code paths) actually JIT compile.